### PR TITLE
test-buildrpm

### PR DESCRIPTION
**Pull request by dickmao**
The title for this PR has changed twice already.  I hope this is the final iteration.

In light of the symlink fix that buildrpm caught, I believe it a useful test to dry-run a buildrpm. While the spec file we use is not at all what @dsteuer uses in production, the tom moertl version is a good enough proxy to head off some (unlikely all) unwelcome surprises anethema to downstream distributors.

### CODE REVIEW

**Comment by dickmao** [comment]
> I don't think anyone here is hard on sticking to the current structure of the tarballs.

This I'm afraid is simply untrue.  The tarball contains nontrivial structure (that has been iterated upon over several years) which `package-build` would have to replicate anyway.

**Comment by vspinu** [comment]
If it's true or not remains to be seen. The status quo is clearly sub-optimal. 

> The tarball contains nontrivial structure

Not sure how this is relevant to the problem. 

We will have a new target `pkgbuild` which will create an installable tar.gz with all lisp files at the root and doc, info and etc directories. 

The `install` target will follow a simplified auto-detection procedure as described in #754  which would first `pkgbuild` and then move the entire content to an automatically detected lisp folder. This should eliminate the need for RPM on the user side and we could remove it from the repo.

We could also have a `install-pkg` target which would run emacs and install the pacakge into local ELPA folder.

File: .travis.yml [hunk 1]
```diff
@@ -34,5 +35,6 @@ before_script:
 script:
 - emacs --version
 - R --version
-- make all
+- make
 - make test
+- make test-buildrpm
```
Comment (jabranham) [inline .travis.yml]: Needs a newline

File: .travis.yml [hunk 2]
```diff
@@ -34,5 +35,6 @@ before_script:
 script:
 - emacs --version
 - R --version
-- make all
+- make
```
Comment (jabranham) [inline .travis.yml]: Why this change?

File: Makefile [hunk 3]
```diff
@@ -1,11 +1,10 @@
 ## Top Level Makefile
 
-## Before making changes here, please take a look at Makeconf
 include ./Makeconf
 
-## 'all' is the default target, i.e. 'make' and 'make all' are the same.
+.DEFAULT_GOAL := all
```
Comment (jabranham) [inline Makefile]: Why this change?

File: Makefile [hunk 4]
```diff
@@ -50,70 +49,61 @@ autoloads:
 
 # Create .tgz and .zip files only
 # GNUTAR=gtar make tarballs
-tarballs: $(ESSDIR)
+.PHONY: rel-tarballs
+rel-tarballs: rel-staging
```
Comment (jabranham) [inline Makefile]: Why the rename?

File: Makefile [hunk 5]
```diff
@@ -50,70 +49,61 @@ autoloads:
...
 	@echo "**********************************************************"
 	@echo "** Making distribution of ESS for (pre)release $(ESSVERSION) from $(ESSDIR)/"
-	@echo "** Making pdf and html documentation"
-	cd $(ESSDIR)/doc/ ; $(MAKE) pdf
-	cd $(ESSDIR)/doc/ ; $(MAKE) html
-	@echo "** Creating .tgz file **"
 	test -f $(ESSDIR).tgz && rm -rf $(ESSDIR).tgz || true
+	grep -E 'defvar ess-(version|revision)' $(ESSDIR)/lisp/ess-custom.el
```
Comment (jabranham) [inline Makefile]: What does this grep do?

File: Makefile [hunk 6]
```diff
@@ -50,70 +49,61 @@ autoloads:
...
+	@echo "** Creating .tgz file **"
 	$(GNUTAR) hcvofz $(ESSDIR).tgz $(ESSDIR)
 	@echo "Signing tgz file"
-	$(GPG) -ba -o $(ESSDIR).tgz.sig $(ESSDIR).tgz
+	-$(GPG) -ba -o $(ESSDIR).tgz.sig $(ESSDIR).tgz
```
Comment (jabranham) [inline Makefile]: What's the difference between `-$` and `$`?

File: Makefile [hunk 7]
```diff
@@ -50,70 +49,61 @@ autoloads:
...
+	cd $(ESSDIR)/doc ; $(MAKE) cleanaux
+	cd lisp; $(MAKE) ess-custom.el; $(INSTALL) ess-custom.el ../$(ESSDIR)/lisp/
+	cd lisp; $(MAKE) julia-mode.el; $(INSTALL) julia-mode.el ../$(ESSDIR)/lisp/
 	$(INSTALL) RPM.spec $(ESSDIR)/
 	chmod a-w $(ESSDIR)/lisp/*.el
 	chmod u+w $(ESSDIR)/lisp/ess-site.el $(ESSDIR)/Make* $(ESSDIR)/*/Makefile
 	touch $(ESSDIR)/etc/.IS.RELEASE
 #	# Get (the first 12 hexdigits of) the git version into the release tarball:
-	cut -c 1-12 $(ESSDIR)-git/.git/refs/heads/master > $(ESSDIR)/etc/git-ref
+	if git diff-index --quiet HEAD -- ; then commit=`git stash create | cut -c 1-12` ; else commit=`git rev-parse HEAD | cut -c 1-12` ; fi ; echo $commit > $(ESSDIR)/etc/git-ref
```
Comment (jabranham) [inline Makefile]: This line is too long.

File: Makefile [hunk 8]
```diff
@@ -50,70 +49,61 @@ autoloads:
...
 
-dist: VERSION tarballs
-	grep -E 'defvar ess-(version|revision)' lisp/ess-custom.el \
-	  $(ESSDIR)/lisp/ess-custom.el > $@
+dist:
```
Comment (jabranham) [inline Makefile]: Isn't this the same as `all`?

File: RPM.spec.in [hunk 9]
```diff
@@ -1,14 +1,14 @@
 # This is an RPM spec file that specifies how to package
 # ESS for Fedora Core Linux and, possibly, similar systems.
 
-%define name emacs-ess
```
Comment (jabranham) [inline RPM.spec.in]: Why the name change?

File: Makefile [hunk 10]
```diff
@@ -165,24 +155,25 @@ upload:
...
 ##     by the down stream maintainers:
 .PHONY: buildrpm
-buildrpm: dist
+buildrpm: rel-tarballs
 	rpmbuild -ta --sign $(ESSDIR).tgz
 
-builddeb:
-	dpkg-buildpackage -uc -us -rfakeroot -tc
+.PHONY: test-buildrpm
+test-buildrpm: rel-tarballs
```
Comment (jabranham) [inline Makefile]: Why is this .PHONY? Shouldn't it produce a file?

**Comment by jabranham** [comment]
I'm still opposed to ESS trying to package itself for a specific distro. What happens when someone submits a PR that breaks this test? No one on the development team knows about RPM packaging.

File: RPM.spec.in (hunk 9 above)
Comment (dickmao) [reply RPM.spec.in]: [we good now](https://github.com/emacs-ess/ESS/pull/732#issuecomment-439958230)

File: Makefile (hunk 10 above)
Comment (dickmao) [reply Makefile]: [we good now](https://github.com/emacs-ess/ESS/pull/732#issuecomment-439958230)

File: Makefile (hunk 7 above)
Comment (dickmao) [reply Makefile]: done.

File: Makefile (hunk 6 above)
Comment (dickmao) [reply Makefile]: Prevent gpg from failing the recipe.
Travis doesn't have a gpg key, and I don't want travis to fail.

File: Makefile (hunk 5 above)
Comment (dickmao) [reply Makefile]: You might have to ask [mmaechler](https://github.com/emacs-ess/ESS/blob/master/Makefile#L125) but I think #753 obviates it.  Removing with extreme prejudice.

File: Makefile (hunk 4 above)
Comment (dickmao) [reply Makefile]: `rel` is the umbrella make directive by which mmaechler creates the self-contained tarball.  I think it's important to associate the `tarballs` target with `rel`.

File: Makefile (hunk 3 above)
Comment (dickmao) [reply Makefile]: I generally don't like to depend on the target that lexicaly first appears being the default target.  I like to be explicit about the default target.

File: .travis.yml (hunk 2 above)
Comment (dickmao) [reply .travis.yml]: Went back to `make all`

File: .travis.yml (hunk 1 above)
Comment (dickmao) [reply .travis.yml]: done

File: Makefile (hunk 8 above)
Comment (dickmao) [reply Makefile]: yes!  reworked it.

**Comment by dickmao** [comment]
[dsteuer](https://github.com/emacs-ess/ESS/pull/732#issuecomment-439958230) has confirmed our initial suspicion that the make machinery for rpm is unused.  This PR is now back to subsuming #707 .

File: Makefile (hunk 3 above)
Comment (vspinu) [reply Makefile]: Well, this is such a commonality that that line is no more than unnecessary noise. It's super easy to detect mistakes in default target.

File: Makefile (hunk 5 above)
Comment (vspinu) [reply Makefile]: versions are not longer in ess-custom but in ess.el and that grep is almost surely not needed.

File: Makefile (hunk 6 above)
Comment (vspinu) [reply Makefile]: But that means the local GPG error can be missed as well, right?

File: Makefile [hunk 11]
```diff
@@ -72,71 +71,57 @@ essr: VERSION
...
 	chmod a-w $(ESSDIR)/lisp/*.el
 	chmod u+w $(ESSDIR)/lisp/ess-site.el $(ESSDIR)/Make* $(ESSDIR)/*/Makefile
 	touch $(ESSDIR)/etc/.IS.RELEASE
 #	# Get (the first 12 hexdigits of) the git version into the release tarball:
-	cut -c 1-12 $(ESSDIR)-git/.git/refs/heads/master > $(ESSDIR)/etc/git-ref
-
-dist: VERSION tarballs
-	grep -E 'defvar ess-(version|revision)' lisp/ess-custom.el \
-	  $(ESSDIR)/lisp/ess-custom.el > $@
+	if git diff-index --quiet HEAD -- ; then \
```
Comment (vspinu) [inline Makefile]: Could you please describe for mortals what this one is doing and why it is needed?

Comment (dickmao) [reply Makefile]: You git clone master.  You change VERSION.  You run `make tarballs` to see if it takes.

It doesn't because [you're still getting the Makefile from master](https://github.com/emacs-ess/ESS/blob/47bd3d257f99eec427d296226826e1c1b45919bd/Makefile#L107).

The line in question says: "If you have made a change in your working tree, create a new git-ref and pack it into the tarball.  If you have not made any changes, pack the master's git-ref."

File: Makefile (hunk 6 above)
Comment (dickmao) [reply Makefile]: Fixed

File: Makefile [hunk 12]
```diff
@@ -72,71 +72,61 @@ essr: VERSION
...
-#	the making of pdf, html, ESSR.rds shall go into `make dist' #752
-	@cd $(ESSDIR)/doc/ ; $(MAKE) pdf
-	@cd $(ESSDIR)/doc/ ; $(MAKE) html
-	@echo "** Creating .tgz file **"
+	@echo "** Making distribution of ESS for (pre)release $(ESSVERSION) from $(ESSDIR)/"
 	test -f $(ESSDIR).tgz && rm -rf $(ESSDIR).tgz || true
+	@echo "** Creating .tgz file **"
 	$(GNUTAR) hcvofz $(ESSDIR).tgz $(ESSDIR)
 	@echo "Signing tgz file"
+ifeq ($(TRAVIS),)
```
Comment (jabranham) [inline Makefile]: We use `CONTINUOUS_INTEGRATION` elsewhere for this sort of thing, might as well use it here for consistency.

File: lisp/Makefile [hunk 13]
```diff
@@ -38,7 +38,7 @@ install: dist
 	$(INSTALL) $(ELC) $(LISPDIR)
 	if [ -f /etc/debian_version -a -n "$(SITELISP)" -a ! -f "$(SITELISP)/ess-site.el" ] ; \
 	then \
-		ln -s $(LISPDIR)/ess-site.el $(SITELISP)/ess-site.el ; \
+		cd $(SITELISP) ; ln -s ess/ess-site.el . ; \
```
Comment (jabranham) [inline lisp/Makefile]: Why this change?

File: lisp/Makefile [hunk 14]
```diff
@@ -51,6 +51,11 @@ uninstall:
 distclean clean:
 	rm -f $(ELC) ess-autoloads.el julia-mode.el .dependencies
 
+.PHONY: test
+test:
+	! ( $(COMPILE) $(shell git ls-files *.el) 2>&1 | awk '{if (/^ /) { gsub(/^ +/, " ", $$0); printf "%s", $$0 } else { printf "\n%s", $$0 }}' | egrep -a "not known|Error|free variable|error for|Use of gv-ref" )
```
Comment (jabranham) [inline lisp/Makefile]: Why this new target?

Also, this line is way too long. Please cut lines around 80ish characters.

File: lisp/ess-r-flymake.el [hunk 15]
```diff
@@ -163,9 +163,13 @@ into a list and call REPORT-FN on it."
        for msg = (match-string 4)
        for (beg . end) = (let ((line (string-to-number (match-string 1)))
                                (col (string-to-number (match-string 2))))
-                           (flymake-diag-region src-buffer line col))
+                           (if (fboundp 'flymake-diag-region)
```
Comment (jabranham) [inline lisp/ess-r-flymake.el]: This change isn't necessary. This file assumes Emacs 26+

File: lisp/ess-r-mode.el [hunk 16]
```diff
@@ -2293,7 +2293,8 @@ state.")
             (end (match-end 0))
             ;; Remove surrounding quotes so the text property is the
             ;; name of the link:
-            (text (string-trim (match-string 0) "[\"‘]*" "[\"’]*")))
+            (text (replace-regexp-in-string "^[\"']+" ""
```
Comment (jabranham) [inline lisp/ess-r-mode.el]: I've already fixed this in master. Please don't include changes that aren't related to the PR, it makes it harder to review.

File: etc/Makefile [hunk 17]
```diff
@@ -23,6 +23,9 @@ ESSR_CODE_FILES = ESSR/R/*.R ESSR/R/.*.R
 
 all:
 
+.PHONY: dist
+dist: all
```
Comment (jabranham) [inline etc/Makefile]: I don't think we need this.

**Comment by jabranham** [comment]
Can't we just add `make tarballs` to the travis file as it is now and let the gpg signing fail on travis? That seems much simpler than this PR.

File: lisp/Makefile (hunk 13 above)
Comment (dickmao) [reply lisp/Makefile]: This actually came out of the rpm testing.
`buildrpm` (correctly!) complained about the symlink being absolute.  It makes a lot of sense that symlinks in the tarball "microcosm" ought to be relative to avoid assuming anything about where the files ultimately end up.

File: etc/Makefile (hunk 17 above)
Comment (dickmao) [reply etc/Makefile]: You will see a `make dist` in the `rel-staging` target in the parent Makefile.  It calls `make dist` for `lisp`, `etc` and `doc` (thus resolving the pdfs and html).

File: lisp/ess-r-flymake.el (hunk 15 above)
Comment (dickmao) [reply lisp/ess-r-flymake.el]: But the compilation test fails without it.

File: lisp/Makefile (hunk 14 above)
Comment (dickmao) [reply lisp/Makefile]: I almost have to insist on the lisp compiling cleanly.  It found the error in #748

Comment (jabranham) [reply lisp/Makefile]: This isn't the right way to test for byte compilation warnings. You'd want to set `byte-compile-error-on-warn` to `t` while compiling.

We're already doing that for Emacs 26+ on travis. I was the one responsible for fixing several hundred byte compiler warnings and I don't use Emacs 25 but we should probably enable it there too. I'll look into grabbing an old Emacs and seeing what needs fixing.

Comment (dickmao) [reply lisp/Makefile]: ah, okay.  i knew there had to be a better way.

**Comment by dickmao** [comment]
> [dsteuer](https://github.com/emacs-ess/ESS/pull/732#issuecomment-439958230) has confirmed our initial suspicion that the make machinery for rpm is unused. This PR is now back to subsuming #707 .

And we're back to conflicting with #707 again.  One of these days I'm going to add an actual  feature to ESS instead of this ongoing struggle with the plumbing.

In light of the symlink fix that `buildrpm` caught, I believe it a useful test to dry-run a `buildrpm`.  While the spec file we use is not at all what @dsteuer uses in production, the tom moertl version is a good enough proxy to head off the unwelcome surprises anethema to downstream distributors sh.

File: RPM.spec.in (hunk 9 above)
Comment (dickmao) [reply RPM.spec.in]: It was looking for files in `emacs-ess` instead of `ess` which is where they were.  It's not too important as we are merely using RPM.spec as a coarse-grained test.

File: Makefile (hunk 10 above)
Comment (dickmao) [reply Makefile]: I believe all tests should be PHONY targets.

File: Makefile (hunk 12 above)
Comment (dickmao) [reply Makefile]: done
